File: admin/tests/parser-lab-parity.test.mjs
// F-014 — Parser Lab parity, compiled-artifact side of the contract.
//
// Loads the ACTUAL public/parser_lab.js the Lab page executes (a dart2js build
// of the device engine) and proves it reproduces the same goldens the Dart
// engine test (app/test/engine/parser_lab_parity_test.dart) asserts. A stale
// or divergent compiled artifact fails HERE while the Dart side stays green —
// the exact F-014 failure mode, made visible in CI instead of in production.

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";
import vm from "node:vm";

const here = dirname(fileURLToPath(import.meta.url));
const fixtures = JSON.parse(
  readFileSync(join(here, "fixtures/parser_parity_fixtures.json"), "utf8"),
);
const goldens = JSON.parse(
  readFileSync(join(here, "fixtures/parser_parity_goldens.json"), "utf8"),
);

// dart2js targets a browser-ish global; give it `self` in a fresh VM context.
const context = vm.createContext({});
context.self = context;
context.globalThis = context;
vm.runInContext(
  readFileSync(join(here, "../public/parser_lab.js"), "utf8"),
  context,
  { filename: "parser_lab.js" },
);

test("compiled parser_lab.js exposes the rules-aware entry", () => {
  assert.equal(typeof context.parseSmsWithRules, "function");
  assert.equal(typeof context.parseSms, "function");
  assert.ok(String(context.parserLabContract).includes("catalog-rules"),
    "compiled artifact predates the catalog-rules contract — recompile");
});

test("compiled engine reproduces the shared parity goldens", () => {
  const rulesJson = JSON.stringify(fixtures.rules);
  fixtures.cases.forEach((c, i) => {
    const g = goldens[i];
    const raw = context.parseSmsWithRules(c.body, c.sender, rulesJson);
    const r = JSON.parse(raw);
    const why = `case ${c.name}: compiled artifact diverges from the Dart ` +
      `engine goldens — recompile parser_lab.js from current app source`;
    assert.equal(r.isTransaction, g.isTransaction, why);
    assert.equal(r.catalogRuleId ?? null, g.catalogRuleId, why);
    assert.equal(r.amount ?? null, g.amount, why);
    assert.equal(r.currency ?? null, g.currency, why);
    assert.equal(r.type ?? null, g.type, why);
    assert.equal(r.merchant ?? null, g.merchant, why);
    assert.equal(r.balanceAfter ?? null, g.balanceAfter, why);
    assert.equal(r.parseConfidence ?? null, g.parseConfidence, why);
  });
});

File: app/test/engine/parser_lab_parity_test.dart
// F-014 — Parser Lab parity, device side of the contract.
//
// The lab runs `admin/public/parser_lab.js`, a dart2js compilation of THIS
// engine. Parity holds when both sides produce the goldens below for the
// shared fixtures. The Dart engine (this test) is the AUTHORITY: when engine
// behaviour changes intentionally, regenerate the goldens from the engine and
// recompile parser_lab.js — `admin/tests/parser-lab-parity.test.mjs` then
// proves the compiled artifact is not stale. Neither side ever copies the
// other's output at runtime; the goldens are the meeting point.

import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:money_companion/engine/parser/bank_profile.dart';
import 'package:money_companion/engine/parser/catalog_rule_matcher.dart';
import 'package:money_companion/engine/parser/parser_engine.dart';

void main() {
  test('F-014: the Dart engine reproduces the shared parity goldens', () {
    const dir = '../admin/tests/fixtures';
    final fixtures = jsonDecode(
        File('$dir/parser_parity_fixtures.json').readAsStringSync()) as Map;
    final goldens = jsonDecode(
        File('$dir/parser_parity_goldens.json').readAsStringSync()) as List;

    final rules = [
      for (final r in fixtures['rules'] as List)
        CatalogParserRule(
          id: r['id'] as String,
          senderPattern: r['sender_pattern'] as String,
          messagePattern: r['message_pattern'] as String,
          transactionType: r['transaction_type'] as String,
          priority: r['priority'] as int,
          extractedFields: (r['extracted_fields'] as Map)
              .map((k, v) => MapEntry(k.toString(), v)),
        ),
    ];

    const engine = ParserEngine();
    final cases = fixtures['cases'] as List;
    expect(goldens.length, cases.length,
        reason: 'goldens out of sync with fixtures — regenerate');

    for (var i = 0; i < cases.length; i++) {
      final c = cases[i] as Map;
      final g = goldens[i] as Map;
      final res = engine.parse(
        c['body'] as String,
        senderId: c['sender'] as String,
        bankProfiles: BankProfiles.all,
        catalogRules: rules,
      );
      final reason = 'case ${c['name']}: engine drifted from goldens — if '
          'intentional, regenerate goldens AND recompile parser_lab.js';
      expect(res.isTransaction, g['isTransaction'], reason: reason);
      expect(res.catalogRuleId, g['catalogRuleId'], reason: reason);
      expect(res.transaction?.amount, g['amount'], reason: reason);
      expect(res.transaction?.currency, g['currency'], reason: reason);
      expect(res.transaction?.type.name, g['type'], reason: reason);
      expect(res.transaction?.rawMerchant, g['merchant'], reason: reason);
      expect(res.transaction?.balanceAfter, g['balanceAfter'], reason: reason);
      expect(res.transaction?.parseConfidence, g['parseConfidence'],
          reason: reason);
    }
  });
}
